graph_algorithms/algorithms/examples/louvain.gsql

```diff
@@ -1,80 +1,295 @@
+                PRINT iteration AS Phase1Iter, @@modularity;
+                log(debug > 0, "[redrain]#move", iteration, @@modularity);
         END;
 
-        Start = {ANY};
+# Phase 2 --  Merge     
+        iteration2 = 0;
+        WHILE (iteration2 < 2 OR @@modularity2 - last_modularity2 > epsilon) LIMIT iter2 DO
+                iteration2 += 1;
+                Start = SELECT s
+                        FROM Start:s
+                        ACCUM s.@uid = s.@cid;      
+                # Select the vertices with minimal internal id to represent the coarsened graph
+                Start = SELECT s
+                        FROM Start:s 
+                        ACCUM @@representMap += (s.@cid -> s);
+        
+                FOREACH (key, value) IN @@representMap DO
+                        @@representSet += value;                       
+                END;      
+                represent = {@@representSet};
+                @@representMap.clear();
+                @@representSet.clear();
+                log(debug > 0, "[redrain]#2_merge", represent.size()); #@@clusterSizes.size());
 
+            # Get @cweight from totalIncident
+                represent = SELECT s
+                            FROM represent:s
+                            ACCUM s.@cweight = @@totIncidentCluster.get(s.@uid),
+                                  @@clusterSizes += (s.@cid -> 1);
 
+                log(debug > 1, "[redrain]#2_merge", @@weightToClusterMap.size());
+                iteration = 0;
+                last_modularity = 0;
+                @@modularity = 0;
+     
+                WHILE (iteration < 2 OR @@modularity - last_modularity > epsilon) limit iter1 DO
+                        iteration += 1;
+        
+                        # Calculate.weight incident from vertex to cluster in coarsened graph; change every interation
+                        S = SELECT s
+                            FROM Start:s -(Friend:e)-:t
+                            WHERE s.@cid != t.@cid AND @@totIncidentCluster.get(s.@uid) > 0 AND @@totIncidentCluster.get(t.@cid) > 0   #@@totIncidentCluster keeps changing, can be 0
+                            ACCUM @@weightToClusterMap += (s.@uid -> (t.@cid -> e.weight));  # from s, incident to some clusters. Not consider the same cluster
+                        represent = SELECT s 
+                                    FROM represent:s
+                                    POST-ACCUM INT bestCluster = s.@cid,
+                                               FLOAT maxDeltaQ = 0.0,
+                                               FLOAT deltaQ_new = 0.0,
+                                               FOREACH (cluster, weightToC) IN @@weightToClusterMap.get(s.@uid) DO 
+                                                       FLOAT incident = @@totIncidentCluster.get(cluster),
+                                                       IF @@clusterSizes.get(s.@cid) == 1 AND @@clusterSizes.get(cluster) == 1 AND s.@cid < cluster THEN
```

Comment:
Why the outdegree of a vertex can be checked by weight map and @weight, but it creates a separate map for cluster merging? Can this be checked through weight?

Reply:
Because here we can only check s.@cweight and the weight for the cluster with label s.@cid, but we can't check the @weight for the vertices in the cluster which we want to move to.